### PR TITLE
RFC 002

### PR DESCRIPTION
**Pull request by flametail9**
Reference counting, strides, transpose, and sub-matrix support.

### CODE REVIEW

File: projects/matlib/Mat.h [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+        }
+        size_type rows() const{
+            return this->dims[0];
+        }
+        size_type columns() const{
+            return this->dims[1];
+        }
+        size_type size() const{
+            return rows()*columns();
+        }
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:49]: move up to be with iterator stuff

File: projects/matlib/Mat.h [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        Mat(size_type a = 1, size_type b = 1){
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:50]: Eventually make variadic

File: projects/matlib/Mat.h [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+            dims[0] = a;
+            dims[1] = b;
+            strides = new size_type[ndims];
+            strides[0] = 1;
+            strides[1] = columns();
+            memory = new Type[a*b];
+            data = memory;
+            size_type i = 0;
+            for(auto elem : list){
+                errorCheck(i >= size(), "Initializer out of bounds\n");
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:75]: prob check if too small as well

File: projects/matlib/Mat.h [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+            errorCheck(*refCount < 0,"Reference counter is negative somehow\n");
+            if(*refCount == 0){
+                delete refCount;
+                delete []memory;
+            }
+            delete []dims;
+            delete []strides;
+        }
+        Type& operator() (size_type a, size_type b){
+            errorCheck(a < 0 || b < 0 || a + b > size(),"Element access outside matrix scope\n");
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:103]: derp

File: projects/matlib/Mat.h [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+            dims = new size_type[ndims];
+            dims[0] = b.dims[0];
+            dims[1] = b.dims[1];
+            strides = new size_type[ndims];
+            strides[0] = b.strides[0];
+            strides[1] = b.strides[1];
+            return *this;
+        }
+        Mat operator+ (const Mat &b){
+            size_type i = 0;
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:152]: might be redundant?

File: projects/matlib/Mat.h [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+                if(dims[n] == 1) x[i] = b.dims[i];
+                else x[i] = dims[i];
+                i++;
+            }
+            Mat<Type> result(x[0], x[1]);
+            delete []x;
+
+            size_type width;
+            if(columns() > b.columns()) width = columns();
+            else width = b.columns();
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:168]: = result.columns()

File: projects/matlib/Mat.h [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+            }
+            return result;
+        }
+        Mat operator- (){
+            Mat<Type> result(rows(),columns());
+            size_type n = 0;
+            for(auto i : *this){
+                result(0,n) = i * -1;
+                n++;
+            }
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:203]: try 
```
for(auto& i : *this){
    i *= -1;
}
```

File: projects/matlib/Mat.h [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+                }
+            }
+            return result;
+        }
+        void T(Mat& dest){
+            this->errorCheck(this->size() != dest.size(),"Matrix size mismatch");
+            this->errorCheck(this->data == dest.data, "Source and destination matrix share same backing data");
+            for(size_t i=0;i<this->size();i++){
+                dest(0,i) = this->operator()(i%this->rows(),i/this->rows());
+            }
```
Comment (anuranbaka) [inline projects/matlib/Mat.h]: Perhaps require that matrix is continuous? Check continuous first and then do it element wise with a double loop if not contiguous.

File: projects/matlib/Mat.h [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+                        sum += this->operator()(x,n)*b(n,i);
+                    }
+                    result(x,i) = sum;
+                }
+            }
+            return result;
+        }
+        void T(Mat& dest){
+            this->errorCheck(this->size() != dest.size(),"Matrix size mismatch");
+            this->errorCheck(this->data == dest.data, "Source and destination matrix share same backing data");
```
Comment (anuranbaka) [inline projects/matlib/Mat.h]: check memory, not data

Comment (anuranbaka) [reply projects/matlib/Mat.h]: Put a TODO error to call the other transpose if same data pointer and appropriate strides+shape output

File: projects/matlib/Mat.h [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                    sum = 0;
+                    for(size_t n = 0; n< this->columns(); n++){
+                        sum += this->operator()(x,n)*b(n,i);
+                    }
+                    result(x,i) = sum;
+                }
+            }
+            return result;
+        }
+        void T(Mat& dest){
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:224]: check that ouput shape is compatible

File: projects/matlib/Mat.h [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+            this->errorCheck(this->data == dest.data, "Source and destination matrix share same backing data");
+            for(size_t i=0;i<this->size();i++){
+                dest(0,i) = this->operator()(i%this->rows(),i/this->rows());
+            }
+            size_t temp = dest.dims[0];
+            if(this->columns() != dest.rows()) dest.dims[0] = dest.dims[1];
+            if(this->rows() != dest.columns()) dest.dims[1] = temp;
+            return;
+        }
+        Mat T(){
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:235]: swap dims

File: projects/matlib/Mat.h [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        }
+        void T(Mat& dest){
+            this->errorCheck(this->size() != dest.size(),"Matrix size mismatch");
+            this->errorCheck(this->data == dest.data, "Source and destination matrix share same backing data");
+            for(size_t i=0;i<this->size();i++){
+                dest(0,i) = this->operator()(i%this->rows(),i/this->rows());
+            }
+            size_t temp = dest.dims[0];
+            if(this->columns() != dest.rows()) dest.dims[0] = dest.dims[1];
+            if(this->rows() != dest.columns()) dest.dims[1] = temp;
```
Comment (anuranbaka) [inline projects/matlib/Mat.h]: don't mess with destination shape, just error if the shapes are incompatible

File: projects/matlib/Mat.h [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+            return;
+        }
+        Mat T(){
+            Mat<Type> dest(this->columns(),this->rows());
+            for(size_t i=0;i<this->size();i++){
```
Comment (anuranbaka) [inline projects/matlib/Mat.h]: non contiguous agian

File: projects/matlib/Mat.h [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+            matrix.errorCheck(matrix.data != b.matrix.data, "Comparison between iterators of different matrices");
+            if(position == b.position) return true;
+            else return false;
+        }
+        bool operator!=(MatIter b){
+            matrix.errorCheck(matrix.data != b.matrix.data, "Comparison between iterators of different matrices");
+            if(position != b.position) return true;
+            else return false;
+        }
+        MatIter& operator++(){
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:314]: Test iteration on 1x1

Comment (anuranbaka) [reply projects/matlib/Mat.h:314]: test 0x0

File: projects/matlib/Mat.h (hunk 9 above)
Comment (flametail9) [reply projects/matlib/Mat.h]: Not sure what you mean by a TODO error. Shouldn't I just put:
`T(); return;`
?

File: projects/matlib/Mat.h (hunk 14 above)
Comment (flametail9) [reply projects/matlib/Mat.h:314]: Works as intended! :D

File: projects/matlib/Mat.h [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+            errorCheck(data == dest.data, "Source and destination matrix share same backing data\n");
+            for(size_type i=0;i<size();i++){
+                dest(0,i) = operator()(i%rows(),i/rows());
+            }
+            size_type temp = dest.dims[0];
+            if(columns() != dest.rows()) dest.dims[0] = dest.dims[1];
+            if(rows() != dest.columns()) dest.dims[1] = temp;
+            return;
+        }
+        Mat T(){
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:235]: check for non square matrix in matrix

**Pull request merged**